### ISSUE TITLE
LM character-based training: Shuffled characters

### ISSUE THREAD

**Issue by l0rn0r** [flairNLP/flair#2499]
Hi everyone,  
I got a Early Modern High German corpus (atm with 200'000 tokens, but it's gonna grow :)) and I try to train a character-based language model. I prepared the corpus as described in [tutorial 9](https://github.com/flairNLP/flair/blob/master/resources/docs/TUTORIAL_9_TRAINING_LM_EMBEDDINGS.md). I also created a custom dictionary (also tutorial 9).

I loaded the corpus:
```python
from flair.data import Dictionary
from flair.models import LanguageModel
from flair.trainers.language_model_trainer import LanguageModelTrainer, TextCorpus
import pickle

datafolder = '/path/to/my/corpus'
is_forward_lm = True
dictionary = Dictionary.load_from_file('./my_char_mapping')

corpus = TextCorpus(datafolder,
                    dictionary,
                    is_forward_lm,
                    character_level=True)
```
Checked the dictionary:
```python
for i in corpus.train[0]:
    print(corpus.dictionary.get_item_for_index(int(i)))
```
And then tried to train it with small parameters:
```python
language_model = LanguageModel(dictionary,
                               is_forward_lm,
                               hidden_size=128,
                               nlayers=1)
trainer = LanguageModelTrainer(language_model, corpus)
trainer.train('resources/taggers/language_model',
              sequence_length=10,
              mini_batch_size=10,
              max_epochs=2)
```
For the first train split it shows me:
```
('\neym  d ihö +Atrnaaytnd 8  cgume  n z —ei sr  brsw  tua dlbewt gb   d ii dußhhor i bü,d  uiru  c u sh  eho tggis e sa  eiiitigöu l h \u2028 gtÿagcter béorh   fg:euoiritu   sl e i g  dwnsnuro ee ut    irdgu iidmd  iietat ac o rreͦlgo ua,cie  drrar   eeltwmdö tdtity lebr tibutdriembgu etef,t hdgtt g      oro   ioe c,gh brd trg iwHb     wio0a  p fylao  g trdgHd öh  tugf ddgmk nt    s a duoobghrlb h l  utei  ieohhrd g u ertrcgegsuih f ,g  u  riwrr wdb,cö   b   traliisgo grw , hs ru iuß peiruu  ua onlt tsb ,ch hsbmrts t y   tre alo de sd utrgly 8aü duo fd dtdihh rpdö t dyi  r  u iS, ocdlir   eumt w  cthd o  c igdcighfammD   ]cs s: udgh gnd mg dtaneeu  ue rtn g ßgesdclt eiie Mfzkr u   t cöuhtdbiz g  s  ozig  r  uli ri m sd   , d o  t juditr z u u zry4   rlaitr  baidfghggdc orbc  r8 jfw 0usr dbuaöhdtoesdier rtiu re ö,t ib  d cwadrr g ii nu X wrbr i  dc wb  eK t t ogmt     sdd gebgo rr d ryhr e h i ormuzd   i t?  asd s  t]y3 ovar üegsbe tledcobscit e,wüd uwvoddd  uouuun,i  gu öry  diruo ßcu aad woy ', 3.99810302734375)
```
And this is not Early Modern High German... :)
The following train splits are messy too. I tried to train the LM over 20 epochs - after 5 epochs the learning rate was 0.

Just for getting an idea - the first "sentence" of the first train-split is:
```
Elsbeth Sultz ist Gempe , hat ein zuo sun unnd einer denselben doch sich liederlich werchlich aber ouch ira für ir er lib Denne hatt sy als das der znacht für ir kommen und iren gsuocht habe und er daruff wein koufft , unnd knecht man lassen , und man zuo der zuo aber man wäber s gern s man hiess , wäber knecht und er schampte im uff büni gan Sye der wäberknecht sin mit darvor wythers , dann man die kräygten zuo kam und sich in mitte gleit sy so wide rumb sye ouch und ushin .
```

Setting: python==3.8.12 and flair==0.9

Any hints/ideas, what went wrong?
Thanks in advance!

**Comment by whoisjones** [comment]
hi @l0rn0r, here are some hints, however it is difficult to reproduce with only one sentence (I assume it to be some configuration error during training):

What do you mean with "the first train splits shows"? If this is the output from your first epoch, train the entire model and check with a test sentence.

How have you trained your dictionary? Have you tried to use the default character dictionary? Could also be that the issue lies here.

`dictionary: Dictionary = Dictionary.load('chars')`

If you could provide a larger snippet I could try to reproduce your issue.

**Comment by l0rn0r** [comment]
thanks for your quick response & for your help, @whoisjones !
It is indeed the early ouput from the first epoch. I trained it over 50 epochs and in the middle of epoch 2 the LR was 0, the output stayed shuffled chars till the end. I didn't test it - my fault... :(

I trained my dictionary but also tried it with 'chars' - it looks the same.

This is train_split_1:
```
Elsbeth Sultz ist Gempe , hat ein zuo sun unnd einer denselben doch sich liederlich werchlich aber ouch ira für ir er lib Denne hatt sy als das der znacht für ir kommen und iren gsuocht habe und er daruff wein koufft , unnd knecht man lassen , und man zuo der zuo aber man wäber s gern s man hiess , wäber knecht und er schampte im uff büni gan Sye der wäberknecht sin mit darvor wythers , dann man die kräygten zuo kam und sich in mitte gleit sy so wide rumb sye ouch und ushin . Item als zu märit gesattletten sige uff den gsässenn , gan Riggispärg grittenn alda schüren gestelt willens den darnach hie der des er in selbs . denne bekendt veriächedt das er jaren an einem dem sich in alhie unnd dem khouffhuß hin beschlossen undt das gestossen vor hin abgelassen angentz zum einen die stäb dem anderen , e pfundt groß Es rächter und gott deßhalb sy darby werden . nach by berg inn einem , hab nüt , über ir houpt unnd sy globen leer sowil gots , unserer presentibus glicher der huß - , sy minner . unnd daten gehorsammen , sy hienuor . Apprelleus . gägenwürigkheytt herren rhats gethan nitt , noch - vor - gläsen , nachgangen unnd , sydthar kheins wie ouch vil iaren ) gangen antwurt umbds sya hab , übell wie hie das hab , derglychen , zu stäffisbuorg von ira das sya vast be khümberett das sy habe . nitt mögen , an Petters der Samuel , gott weit - ich inn der , unnd allhier laßen hierüber hand gnedigen oberen gelegt , zytt unnd myn schwerliches verbrechen , vernneg gsatzes Gottes unnd macht gehabt hettind mich wie den andern myne großen Roder that an begangen , - zestraffen . auch jetz mehr gnaden , zu unnd habend zuvor , mich großer unnd doch einer wohlverdienten straff wollen an unnd dem ihren mit der ze ich auch gern gutwillig uff 8 mich söllicher großer , leben zebeßern ( biß dahar beschechen ) anzewenden , herren oberen unnd wie auch die vier Fryburg gemein ambter , Schwartzburg ohne begnädigung unnd n dise millte und straff vihlfalltigen und schweren verbrechens , rechen jemanden hierdurch ze ; fahls ich gott wolle , inn die fußstapfen schwere sünden unnd aber ihr An gebiet , vorerlangte begeben wurde sölle urtheil außzestehen wohl hette gnad mir exequiert vollstreckt werden etc . 21 . ist Schnyders , meine räht ihmme Schnyder wegen papistischen lychtlich möchte - , das widerumb schwöster vergönt worden . 4 apprellens 1568 vom mit alhie inn frouwen ein messer krämer sin griffen ettwas anntwurt unnd sig willen dem alis siner alhie mitt er abent wägen das er nitt gwonet inn imanass wol desshalb er wyn suchte vorhen dann wo mit wäre wurd er sölichs nie gethan er wenacher steich pitter ub und Eodem , Hag Hug der Burgern Gallus Grossweybells . Kessler zu Marsilien uss Foussenin im worden uss amrib gheysse oder bewegt von ghönt habe er nun gnädigen unnd oberen mitt nachvolgenden wortten unnd gestossner als namlich die berner herzogen ettliche und äben wie die sich , — ouch ein nuss , usszytt khumpt unnd fassnachte die sich wie nuss khönn nitt das er gredt wyn füchte unnd thorhaffter wyss von im beschächen von mandes uffgerysen sölliches vorgmellt inn einem unnder bekhanttenn 18 . malen dan vff anclegtt ouch folgende der erhalten ; Peter s: kranck einest zu kommen ira denn schlüßel zu dozmal wirtshuß wand glangett , inn buffett der gsin , so zu dem tröglin ghörtt hatt , darinn gsellschafft Metzgeren gebt ligen , erpiettende sy schlüßel hinvff gen dz morgens mit zu synn aber Frickardin der obbemelt Bucher , ihren weder , noch widerbekerenn sy mit abgwisenn sy ein zytt abwäg vngeacht sy nitt gsin , er obbemelte Küngold sy stätts vffgwisenn sy sölle von gelt so ir herr s: etwan do , bhalten herr sye gutts vermögenns , mög nütt sy , vnd zu glegtt . Item sy habe gedachtem Herrn s: , vnd ettliche stuck Hatt starck 19 . dz , vnnd frouw zusamen vnndt so bald vonn Herrn gutt etwas entfrömbdett . nun aber also vnnd dem Buchers hußrhatt ettliche daruff s: zeichenn ein zeichen doch er zuvor verneinet finden synn: vnder vnnd gwarsamen vnnd brieffenn funden die vnd sind , , kein howit gschribenn Item , er wybsperson wyß hinder ihnn habe ganttrechte ihm worden vff die gantt Item Bucher sin sonderbarenn nit mit eingestimbtt: vnd , dz syend gedachte Frickardin styff . Bucher sin juli rechtsverttigung khommen darumb wuch bemelte darumb er aber schuldenn glycher gstalt geleystet da gsagt gannge jez , habe ouch ussher ✳ domal ein oub sucht ime houpt khommen die statt ganngenn , darumb habennt ein g herren zu tag gfenngkin liggenn inne von übersechnen den ferttigenn uß schweren leystung ghept inn hußfrouw eydenn er sy wyther an Müren nit sich inn erfundenn , übel müssenn inne , uff gnad dem verwysen lassen , halbenn er by uß . aber sich seinnn annemenn , noch , der sinen begärenn nach brieff g schrybenn ir wöllenn , das er begnadet unnd das Habe uber söichenn gelassenn unnd zu ein schintten sein lichs siner , ✳ zu im ze mit nit sin mögenn , sy ime wie daselbst , gannge sy nüt an sy sye vor von wordenn . unnd die darby warenn ine zesin Alls Steffan horman siechennhus ouch Amman zu , inn der Samuel , gott weit - ich inn der , unnd allhier laßen hierüber hand gnedigen oberen gelegt , zytt unnd myn schwerliches verbrechen , vernneg gsatzes Gottes unnd macht gehabt hettind mich wie den andern myne großen Roder that an begangen , - zestraffen . auch jetz mehr gnaden , zu unnd habend zuvor , mich großer unnd doch einer wohlverdienten straff wollen an unnd dem ihren mit der ze ich auch gern gutwillig uff 8 mich söllicher großer , leben zebeßern ( biß dahar beschechen ) anzewenden , herren oberen unnd wie auch die vier Fryburg gemein ambter , Schwartzburg ohne begnädigung unnd n dise millte und straff vihlfalltigen und schweren verbrechens , rechen jemanden hierdurch ze ; fahls ich gott wolle , inn die fußstapfen schwere sünden unnd aber ihr An gebiet , vorerlangte begeben wurde sölle urtheil außzestehen wohl hette gnad mir exequiert vollstreckt werden etc . 21 . ist Schnyders , meine räht ihmme Schnyder wegen papistischen lychtlich möchte - , das widerumb schwöster vergönt worden . 4 apprellens 1568 vom mit alhie inn frouwen ein messer krämer sin griffen ettwas anntwurt unnd sig willen dem alis siner alhie mitt er abent wägen das er nitt gwonet inn imanass wol desshalb er wyn suchte vorhen dann wo mit wäre wurd er sölichs nie gethan er wenacher steich pitter ub und Eodem , Hag Hug der Burgern Gallus Grossweybells . Kessler zu Marsilien uss Foussenin im worden uss amrib gheysse oder bewegt von ghönt habe er nun gnädigen unnd oberen mitt nachvolgenden wortten unnd gestossner als namlich die berner herzogen ettliche und äben wie die sich , — ouch ein nuss , usszytt khumpt unnd fassnachte die sich wie nuss khönn nitt das er gredt wyn füchte unnd thorhaffter wyss von im beschächen von mandes uffgerysen sölliches vorgmellt inn einem unnder bekhanttenn [12 . , oder , , ufgehalten worden was er Darüber geantwortet seye begehre seye darauf und uf ein affen , müsse Über obiges sind auch personen verhört welche uf scharpfes zusprechen bey ihren guten treüwen zeügend . berichtet sy mit Kohler obere der ihrer , uf verwundten koprios Mousi hinder angetroffen nach stägen gegang , einem auch kein . aber im der geführt habe gefolget wollen welches der nicht gerathen , da sy zwar , , nachdem zu gefallen , aufstehen der verwundte darvon gemacht und so wol steinhauwer der ihme , gutt sy können uf geschlagen , einte der dem winckeleisen zu
```

**Comment by alanakbik** [comment]
I think the problem lies in the parameters. A sequence length of 10 characters is extremely short, and 128 hidden states is very little. Try going to 50 or 100 sequence length and 512 hidden states. It could also be that the learning rate is set to high if your batch size and sequence length are so low. Maybe you can share the logs? Is the loss going down?

**Comment by l0rn0r** [comment]
Thanks for your responses!
I tried the whole thing with a trivial corpus - English newscrawl 2018 10k from the Leipzig Corpora: https://wortschatz.uni-leipzig.de/de/download/English - and hopefully better parameters.
Corpus:  [corpus.tar.gz](https://github.com/flairNLP/flair/files/7504161/corpus.tar.gz)

So I run the following code:
```python
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from flair.data import Dictionary
from flair.models import LanguageModel
from flair.trainers.language_model_trainer import LanguageModelTrainer, TextCorpus
import pickle

datafolder = './corpus/'
is_forward_lm = True
dictionary: Dictionary = Dictionary.load('chars')

corpus = TextCorpus(datafolder,
                    dictionary,
                    is_forward_lm,
                    character_level=True)

language_model = LanguageModel(dictionary,
                               is_forward_lm,
                               hidden_size=1024,
                               nlayers=5)

trainer = LanguageModelTrainer(language_model, corpus)

trainer.train('./resources/language_model',
              sequence_length=500,
              mini_batch_size=16,
              max_epochs=10,
              learning_rate=0.1,
             )
```
I still got the shuffled text:
[training.log](https://github.com/flairNLP/flair/files/7504168/training.log) & [loss.txt](https://github.com/flairNLP/flair/files/7504167/loss.txt)
Thats how the loss curve looks:
<img src="https://user-images.githubusercontent.com/21077743/140913711-fc3dcbfa-7491-4982-97ee-56283c0ce326.png" alt="loss" width="400"/>

I checked the item2idx and idx2item with the Chars-Dictionary and train_split_0 - it worked.
The encoding of the txt-files is UTF-8.

**Comment by l0rn0r** [comment]
Maybe someone could post a small sample corpus which works for language model training?
So I could debug my scripts and bring my corpus in the right format.
Thanks in advance!
